mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+
+    auto lastArgIndex = fnWithoutMeasurementsBlock->getArguments().size();
+    allocOp.replaceAllUsesWith(fnWithoutMeasurementsBlock->getArgument(lastArgIndex - 1));
+
+    rewriter.eraseOp(allocOp);
+    rewriter.setInsertionPointToStart(&fnWithoutMeasurementsOp.getBody().front());
+
+    std::vector<Operation *> insertOps;
+    fnWithoutMeasurementsOp.walk(
+        [&](quantum::InsertOp insertOp) { insertOps.push_back(insertOp); });
+    fnWithoutMeasurementsOp.walk(
+        [&](func::ReturnOp returnOp) { returnOp->setOperands(insertOps.back()->getResult(0)); });
+
+    quantum::DeallocOp localDealloc = *fnWithoutMeasurementsOp.getOps<quantum::DeallocOp>().begin();
+    rewriter.eraseOp(localDealloc);
+    quantum::removeQuantumMeasurements(fnWithoutMeasurementsOp);
+    return SymbolRefAttr::get(ctx, fnWithoutMeasurementsName);
+}
+FlatSymbolRefAttr ZneLowering::getOrInsertFnWithMeasurements(Location loc,
+                                                             PatternRewriter &rewriter,
+                                                             mitigation::ZneOp op)
+{
+    MLIRContext *ctx = rewriter.getContext();
+    OpBuilder::InsertionGuard guard(rewriter);
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    Type qregType = quantum::QuregType::get(rewriter.getContext());
+    func::FuncOp fnOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+    TypeRange originalTypes = op.getArgs().getTypes();
+
+    std::string fnWithMeasurementsName = op.getCallee().str() + ".withMeasurements";
+    if (moduleOp.lookupSymbol<func::FuncOp>(fnWithMeasurementsName)) {
+        return SymbolRefAttr::get(ctx, fnWithMeasurementsName);
+    }
+    SmallVector<Type> typesWithQreg = {originalTypes.begin(), originalTypes.end()};
+    typesWithQreg.push_back(qregType);
+
+    FunctionType fnWithMeasurementsType = FunctionType::get(ctx, /*inputs=*/
+                                                            typesWithQreg,
+                                                            /*outputs=*/fnOp.getResultTypes());
+    func::FuncOp fnWithMeasurementsOp =
+        rewriter.create<func::FuncOp>(loc, fnWithMeasurementsName, fnWithMeasurementsType);
+    fnWithMeasurementsOp.setPrivate();
+    rewriter.cloneRegionBefore(fnOp.getBody(), fnWithMeasurementsOp.getBody(),
+                               fnWithMeasurementsOp.end());
+    Block *fnWithMeasurementsBlock = &fnWithMeasurementsOp.front();
+    fnWithMeasurementsBlock->addArgument(qregType, loc);
+    quantum::AllocOp allocOpWithMeasurements =
+        *fnWithMeasurementsOp.getOps<quantum::AllocOp>().begin();
+
+    auto lastArgQregIndex = fnWithMeasurementsBlock->getArguments().size();
+    allocOpWithMeasurements.replaceAllUsesWith(
+        fnWithMeasurementsBlock->getArgument(lastArgQregIndex - 1));
+    rewriter.eraseOp(allocOpWithMeasurements);
+    return SymbolRefAttr::get(ctx, fnWithMeasurementsName);
+}
```

Comment:
Thanks for choosing clear names for variables! It helped me a lot to understand the workflow :100: